### ISSUE TITLE
Port twisted.web.static to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6177]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6177|
|Type|enhancement|
|Created|2012-11-08 19:28:52Z|
|Branch|https://github.com/twisted/twisted/tree/twstatic-py3-6177-4|

twisted.web.static should be ported to Python 3.

Attachments:

* [web.static-py3k-wip.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/44b/44bf9611a97b330a7aafa3dda231f3bc7e9b0fbf/cc287f92802172d88ccfeae59f1a89a1fb5ba3c3.patch) (22597 bytes) - added by ihrachyshka on 2013-07-06 14:52:38Z - First draft, not completed

<details><summary>Searchable metadata</summary>

```
trac-id__6177 6177
type__enhancement enhancement
reporter__itamar itamar
priority__lowest lowest
milestone__Python_3_x Python-3.x
branch__branches_twstatic_py3_6177_4 branches/twstatic-py3-6177-4
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1352402932000000 1352402932000000
changetime__1429162405118149 1429162405118149
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ihrachyshka's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ihrachyshka set owner to ihrachyshka<br>ihrachyshka set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ihrachyshka's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|ihrachyshka removed owner<br>ihrachyshka set status to `new`|
|-|-|

I've mostly ported web.static and corresponding unit test to python3.x though it turned out the code depends on python.filepath (thru web.util), while filepath module does not seem to work ok in python3.x (even though corresponding porting ticket was closed as fixed).

For reference:
web.util py3k ticket: [#6178](https://github.com/twisted/twisted/issues/6178)
unicode in filepath: [#5203](https://github.com/twisted/twisted/issues/5203)

**Comment by twisted-trac** [comment]
|<img alt="ihrachyshka's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ihrachyshka commented|
|-|-|

So after applying the patch, run-python3-tests fails while initially importing modules, as follows:

```
Traceback (most recent call last):
  File "./admin/run-python3-tests", line 30, in <module>
    unittest.main(module=None, argv=["run-python3-tests", "-v"] + testModules)
  File "/usr/local/Cellar/python3/3.3.2/Frameworks/Python.framework/Versions/3.3/lib/python3.3/unittest/main.py", line 124, in __init__
    self.parseArgs(argv)
  File "/usr/local/Cellar/python3/3.3.2/Frameworks/Python.framework/Versions/3.3/lib/python3.3/unittest/main.py", line 168, in parseArgs
    self.createTests()
  File "/usr/local/Cellar/python3/3.3.2/Frameworks/Python.framework/Versions/3.3/lib/python3.3/unittest/main.py", line 175, in createTests
    self.module)
  File "/usr/local/Cellar/python3/3.3.2/Frameworks/Python.framework/Versions/3.3/lib/python3.3/unittest/loader.py", line 137, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/local/Cellar/python3/3.3.2/Frameworks/Python.framework/Versions/3.3/lib/python3.3/unittest/loader.py", line 137, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/local/Cellar/python3/3.3.2/Frameworks/Python.framework/Versions/3.3/lib/python3.3/unittest/loader.py", line 96, in loadTestsFromName
    module = __import__('.'.join(parts_copy))
  File "/Users/ihrachyshka/proj/tw/twisted/twisted/web/test/test_static.py", line 19, in <module>
    from twisted.web import static, http, script, resource
  File "/Users/ihrachyshka/proj/tw/twisted/twisted/web/static.py", line 22, in <module>
    from twisted.web.util import redirectTo
  File "/Users/ihrachyshka/proj/tw/twisted/twisted/web/util.py", line 362, in <module>
    class FailureElement(Element):
  File "/Users/ihrachyshka/proj/tw/twisted/twisted/web/util.py", line 372, in FailureElement
    loader = XMLFile(getModule(__name__).filePath.sibling(b"failure.xhtml"))
  File "/Users/ihrachyshka/proj/tw/twisted/twisted/python/filepath.py", line 408, in sibling
    return self.parent().child(path)
  File "/Users/ihrachyshka/proj/tw/twisted/twisted/python/filepath.py", line 681, in child
    newpath = abspath(joinpath(self.path, norm))
  File "/usr/local/Cellar/python3/3.3.2/Frameworks/Python.framework/Versions/3.3/lib/python3.3/posixpath.py", line 92, in join
    "components.") from None
TypeError: Can't mix strings and bytes in path components.
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

(In [44195]) Branching to twstatic-py3-6177.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

(In [44264]) Branching to twstatic-py3-6177-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

(In [44382]) Branching to twstatic-py3-6177-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl removed owner|
|-|-|

Okay. This is a big patch, and I'm sorry!

Some considerations for reviewers:

- This removes the use of Versioned. As it's pickling, this is an acceptable backwards-incompatible change.
- File() handles both *encodings* and *content types* as ```str```.
- This contains an opportunistic port of ```twisted.web.util``` as bits of File requires it.
- Parts that used os.path have been moved to using FilePath for py3 compat reasons.
- Tests that leaked open files have been fixed.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_8"></a>|@wsanchez commented|
|-|-|

In `compat.py:457`: `Return an list` should be `Return a list`

In `test_static.py:435`: `file = resource.openForReading()` could be `with resource.openForReading() as file:` and eliminate the following `file.close()`.  Same at several other points further down in the file.  Also protects against raises in between, though in tests that's probably a minor thing.

Lots to like here, this is awesome even if you don't care about Py3.  Thanks.  Above comments are nits, don't need re-review if you fix them.  You're good to go.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

(In [44481]) Branching to twstatic-py3-6177-4.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44483]) Merge twstatic-py3-6177-4: Port twisted.web.static to Python 3

Author: hawkowl
Reviewer: wsanchez
Fixes: [#6177](https://github.com/twisted/twisted/issues/6177)
